transmat/zip/zip_pack.go

```diff
@@ -0,0 +1,178 @@
+	}
+	defer RequireErrorHasCategory(&err, rio.ErrorCategory(""))
+
+	// Sanitize arguments.
+	if packType != PackType {
+		return api.WareID{}, Errorf(rio.ErrUsage, "this transmat implementation only supports packtype %q (not %q)", PackType, packType)
+	}
+	if !filt.IsComplete() {
+		return api.WareID{}, Errorf(rio.ErrUsage, "filters must be completely specified")
+	}
+	path, err := fs.ParseAbsolutePath(pathStr)
+	if err != nil {
+		return api.WareID{}, Errorf(rio.ErrUsage, "pack must be called with absolute path: %s", err)
+	}
+
+	// Short-circuit exit if the path does not exist.
+	//  We could let the errors later bubble, but, why bother opening a writeController,
+	//  etc, if we're just going to have to rm the resource a millisecond later?
+	afs := osfs.New(path)
+	_, err = afs.Stat(fs.RelPath{})
+	switch Category(err) {
+	case nil:
+		// pass
+	case fs.ErrNotExists:
+		return api.WareID{PackType, ""}, nil
+	default:
+		return api.WareID{}, Errorf(rio.ErrPackInvalid, "cannot read path for packing: %s", err)
+	}
+
+	// Connect to warehouse, and get write controller opened.
+	wc, err := util.OpenWriteController(warehouseAddr, packType, mon)
+	if err != nil {
+		return api.WareID{}, err
+	}
+	defer wc.Close()
+
+	// Construct zip writer.
+	zipWriter := zip.NewWriter(wc)
+
+	// Scan and zip!
+	wareID, err := packZip(ctx, afs, filt, zipWriter)
+	if err != nil {
+		return wareID, err
+	}
+	// Close all the intermediate writer layers to ensure they've flushed.
+	err = zipWriter.Close()
+	if err != nil {
+		return wareID, err
+	}
+
+	// If we made it all the way with no errors, commit.
+	//  (Otherwise, the write controller will be closed by default by our defers.)
+	return wareID, wc.Commit(wareID)
+}
+
+func packZip(
+	ctx context.Context,
+	afs fs.FS,
+	filt api.FilesetPackFilter,
+	zw *zip.Writer,
+) (api.WareID, error) {
+	// Allocate bucket for keeping each metadata entry and content hash;
+	// the full tree hash will be computed from this at the end.
+	bucket := &fshash.MemoryBucket{}
+
+	// Walk the filesystem, emitting entries and filling the bucket as we go.
+	zipHeader := &zip.FileHeader{}
+	preVisit := func(filenode *fs.FilewalkNode) error {
+		if filenode.Err != nil {
+			return filenode.Err
+		}
+
+		// Consider cancellation.
+		if ctx.Err() != nil {
+			return Errorf(rio.ErrCancelled, "cancelled")
+		}
+
+		// Open file.
+		fmeta, file, err := fsOp.ScanFile(afs, filenode.Info.Name) // FIXME : we already have the full metadata loaded; give ScanFile option to accept it!
+		if err != nil {
+			return err
+		}
+
+		// Apply filters.
+		//  The filter may reject things by returning an error;
+		//   or, instruct us to ignore things by setting the type to invalid.
+		if err := filters.ApplyPackFilter(filt, fmeta); err != nil {
+			return err
+		}
+		if fmeta.Type == fs.Type_Invalid {
+			return nil // skip it and continue the walk
+		}
+
+		// Flatten time to seconds.  The tar writer impl doesn't do subsecond precision.
```

Comment:
Comment freshness check ;)

I wonder if the following couple of lines are true of the zip writer as well?